drivers/rtc/rtc_counter.c

```diff
@@ -0,0 +1,471 @@
+	alarm_cfg.callback = counter_rtc_alarm_callback;
+	alarm_cfg.ticks = alarm_ticks;
+	alarm_cfg.user_data = data;
+	alarm_cfg.flags = COUNTER_ALARM_CFG_ABSOLUTE | COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+
+	data->alarm_mask[id] = mask;
+	data->alarm_pending[id] = false;
+
+	return counter_set_channel_alarm(config->counter_dev, (uint8_t)id, &alarm_cfg);
+}
+
+static int counter_rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				      struct rtc_time *timeptr)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(id);
+	ARG_UNUSED(mask);
+	ARG_UNUSED(timeptr);
+
+	return -ENOTSUP;
+}
+
+static int counter_rtc_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	uint32_t key;
+	int ret = 0;
+	struct counter_rtc_data *data = dev->data;
+
+	if (!data->alarm_capable) {
+		return -ENOTSUP;
+	}
+
+	if (id >= data->num_alarm_chans) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+	ret = data->alarm_pending[id] ? 1 : 0;
+	data->alarm_pending[id] = false;
+	irq_unlock(key);
+
+	return ret;
+}
+
+static int counter_rtc_alarm_set_callback(const struct device *dev, uint16_t id,
+					  rtc_alarm_callback callback, void *user_data)
+{
+	uint32_t key;
+	struct counter_rtc_data *data = dev->data;
+
+	if (!data->alarm_capable) {
+		return -ENOTSUP;
+	}
+
+	if (id >= data->num_alarm_chans) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+	data->alarm_callback[id] = callback;
+	data->alarm_user_data[id] = user_data;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+#endif /* CONFIG_RTC_ALARM */
+
+#if defined(CONFIG_RTC_UPDATE)
+
+static int counter_rtc_update_set_callback(const struct device *dev, rtc_update_callback callback,
+					   void *user_data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(callback);
+	ARG_UNUSED(user_data);
+
+	return -ENOTSUP;
+}
+
+#endif /* CONFIG_RTC_UPDATE */
+
+#if defined(CONFIG_RTC_CALIBRATION)
+
+static int counter_rtc_set_calibration(const struct device *dev, int32_t calibration)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(calibration);
+	return -ENOTSUP;
+}
+
+static int counter_rtc_get_calibration(const struct device *dev, int32_t *calibration)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(calibration);
+	return -ENOTSUP;
+}
+
+#endif /* CONFIG_RTC_CALIBRATION */
+
+static int counter_rtc_init(const struct device *dev)
+{
+	const struct counter_rtc_config *config = dev->config;
+	struct counter_rtc_data *data = dev->data;
+	uint32_t freq;
+	uint8_t chans;
+
+	if (!device_is_ready(config->counter_dev)) {
+		LOG_ERR("Counter device %s not ready", config->counter_dev->name);
+		return -ENODEV;
+	}
+
+	/* Require a 1 Hz counter frequency */
+	freq = counter_get_frequency(config->counter_dev);
+
+	if (freq != 1U) {
+		LOG_ERR("Unsupported counter frequency: %u Hz (expected 1 Hz)", freq);
+		return -ENOTSUP;
+	}
+
+	/* Start with zero offset until rtc_set_time is called */
+	data->epoch_offset = 0;
+
+#ifdef CONFIG_RTC_ALARM
+	data->rtc_dev = dev;
+	/* Detect alarm capability from underlying counter */
+	chans = counter_get_num_of_channels(config->counter_dev);
```

Comment:
the rtc-device binding contains alarms-count which shall be set to match the number of channels of the counter, this is known at build time so no need to detect it at runtime